### ISSUE TITLE
Implement logging of downloads with simple DynamoDB on AWS

### ISSUE THREAD

**Issue by Karm** [eclipse/microprofile-starter#28]
Regarding #18 
Concerns mostly:
@Emily-Jiang @cealsair @rdebusscher 

Hi guys, so this is what I suggest as a solution. if it doesn't stir up any opposition I will go ahead and forge this little demo into an application scoped bean callable from [GeneratorDataBean.java](https://github.com/eclipse/microprofile-starter/blob/master/src/main/java/org/eclipse/microprofile/starter/view/GeneratorDataBean.java). Rudy might suggest a better integration though...

# AmazonDynamoDB
We host the service on Amazon, so there is probably no harm in using it for a simple logging of loosely structured data rows. If we want to leave, we always can. Moving away is as easy as dumping that log into JSON and importing it into something else.

# Demo of what we might log
(This is just a showcase, not integrated, it is hard coded, demo, etc.)

```java
package org.eclipse.microprofile.starter;

import com.amazonaws.regions.Regions;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.dynamodbv2.document.PutItemOutcome;
import com.amazonaws.services.dynamodbv2.document.Table;
import org.junit.Test;

import javax.xml.bind.DatatypeConverter;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * @author Michal Karm Babacek <karm@redhat.com>
 */
public class DownloadsLoggerDemoTest {

    @Test
    public void demoLoggingTest() throws NoSuchAlgorithmException {
        final AmazonDynamoDB client = AmazonDynamoDBClientBuilder.standard()
                //.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("http://localhost:8000", "eu-west-1"))
                .withRegion(Regions.EU_WEST_1)
                .build();
        final DynamoDB dynamoDB = new DynamoDB(client);
        try {
            final Table table = dynamoDB.getTable("microprofile_starter_log");

            final Calendar calendar = Calendar.getInstance();
            final SimpleDateFormat formattedTime = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
            final MessageDigest hash = MessageDigest.getInstance("SHA-1");

            // Logged
            final String mpVersion = "2.0";
            final String supportedServer = "liberty";
            final String beansxmlMode = "ANNOTATED";
            final String moduleStructure = "SINGLE";
            final Set<String> selectedSpecs = Stream.of("HEALTH_METRICS", "HEALTH_CHECKS", "REST_CLIENT").collect(Collectors.toSet());
            final String timestamp = formattedTime.format(calendar.getTime());

            // Just used to add entropy to unique stamp
            final String groupId = "biz.kamrs";
            final String artifacId = "demodemo";

            // This should uniquely mark log record row. World wouldn't end if there is a collision though
            final String logmark = Stream.concat(
                    Stream.of(mpVersion, supportedServer, beansxmlMode, moduleStructure, timestamp, groupId, artifacId),
                    selectedSpecs.stream()
            ).collect(Collectors.joining());

            hash.update(logmark.getBytes());

            final String hashedLogmark = DatatypeConverter.printHexBinary(hash.digest());

            System.out.println("logmark:" + logmark);
            System.out.println("logmark hashed:" + hashedLogmark);

            final Item item = new Item()
                    .withPrimaryKey("logmark", hashedLogmark)
                    .withString("mpVersion", mpVersion)
                    .withString("supportedServer", supportedServer)
                    .withString("beansxmlMode", beansxmlMode)
                    .withString("moduleStructure", moduleStructure)
                    .withStringSet("selectedSpecs", selectedSpecs)
                    .withString("timestamp", timestamp);

            PutItemOutcome outcome = table.putItem(item);

            System.out.println("Outcome:" + outcome.toString());

        } finally {
            client.shutdown();
        }
    }
}
```

## Runtime
Once I have AWS_SECRET_ACCESS_KEY and  AWS_ACCESS_KEY_ID env vars set I can run it.  For testing it is possible to use local DynamoDB, but I configured the AWS DynamoDB for our account and tried it on "production" right away:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.microprofile.starter.DownloadsLoggerDemoTest
logmark:2.0libertyANNOTATEDSINGLE2019-01-17T00:37:44.115+0100biz.kamrsdemodemoHEALTH_CHECKSREST_CLIENTHEALTH_METRICS
logmark hashed:6FB59BD3338342E6A34CA39CBDBC73012E021BB7
Outcome:{}
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.38 sec
```

## AWS Console
I executed the "test" twice and what I can see are two log entries:
![dynamo_screenshot](https://user-images.githubusercontent.com/691097/51286407-24ffc500-19f3-11e9-9880-c94493e13f19.png)

## Command line
if you have AWS command line tool installed, you can easily use DynamoDB from your shell, either complex queries or scans, this dumps everything:

```
$ aws dynamodb scan --table-name microprofile_starter_log 
{
    "Count": 2,
    "Items": [
        {
            "moduleStructure": {
                "S": "SINGLE"
            },
            "logmark": {
                "S": "4CBCD59B08945E9C0C84DCE262DE13E700DF34D5"
            },
            "selectedSpecs": {
                "SS": [
                    "HEALTH_CHECKS",
                    "HEALTH_METRICS",
                    "REST_CLIENT"
                ]
            },
            "timestamp": {
                "S": "2019-01-17T00:57:04.348+0100"
            },
            "beansxmlMode": {
                "S": "ANNOTATED"
            },
            "supportedServer": {
                "S": "liberty"
            },
            "mpVersion": {
                "S": "2.0"
            }
        },
        {
            "moduleStructure": {
                "S": "SINGLE"
            },
            "logmark": {
                "S": "6FB59BD3338342E6A34CA39CBDBC73012E021BB7"
            },
            "selectedSpecs": {
                "SS": [
                    "HEALTH_CHECKS",
                    "HEALTH_METRICS",
                    "REST_CLIENT"
                ]
            },
            "timestamp": {
                "S": "2019-01-17T00:37:44.115+0100"
            },
            "beansxmlMode": {
                "S": "ANNOTATED"
            },
            "supportedServer": {
                "S": "liberty"
            },
            "mpVersion": {
                "S": "2.0"
            }
        }
    ],
    "ConsumedCapacity": null,
    "ScannedCount": 2
}
```

# Question

Good / no good?  IMHO it is just what we need for simple logging.

**Comment by rdebusscher** [comment]
@Karm 
I have already the method which creates the JSON and sending it can be done by using a Runnable and the HttpURLConnection and scheduled by the ManagedExecutorService.

Can't we just send a JSON to an endpoint instead of adding the AmazonDynamoDB dependency?

**Comment by Karm** [comment]
@rdebusscher We can do anything we want :smile: 

To use JSON with the aforementioned example, one would simply call:

```
 final Item item = new Item()
                    .withPrimaryKey("logmark", hashedLogmark)
                    .withJSON("log", jsonDocument);
```

Regarding just using HTTP POST somewhere, well, I can whip up a tiny Go service that would talk to the DynamoDB. Or cal teh AWS Dynamo endpoint directly. Note that there would have to be some additional logic on our side with signing headers etc. 
It is true that the DynamoDB AWS SDK is a huge bloatware full of features we don't need and the app is huge enough as it is with JSF.

Lemme take another look on how to write to that store.   I understand that you have no problem with the DynamoDB as such though, right?

**Comment by rdebusscher** [comment]
No problems with DynamoDB as such.

I was just making the remark of changing 10 lines with no additional dependencies with the integration of DynamoDB client.

Hardcoded linking to a certain solution which might change is something the others should also give their feedback on.

But when it is easier to do it this way, fine for me.

**Comment by Karm** [comment]
@rdebusscher It won't be 10 lines, because one has to sign the request header and that is a little laborious, but I have done something like that in the past for Azure storage, [signing requests](https://github.com/Karm/mod_cloud_storage/blob/master/src/main.c#L125).  So I can do it for AWS here too to access DynamoDB. Then one could simply do POST with HttpURLConnection. Ad hard linking to a particular solution - I really don't think it is a problem in any way, because we are not using any of its features, we just drop a line to K:V store.

TL:DR:
 * DynamoDB on AWS: stays :heavy_check_mark:   
 * Using the whole AWS Java SDK: Might be unnecessarily heavy :grey_question: 
 * I will clock some time into preparing manually signed AWS REST putItem request for DynamoDB REST so as we can just use plain HttpURLConnection :heavy_check_mark: 
 * If it proves too weird or fragile, we fall back to AWS Java SDK for the launch :grey_question: 

How about that?

**Comment by Karm** [comment]
@rdebusscher O.K. I have it prototyped, pure HTTP POST, no AWS SDK library.  Gonna flash it out and share here later, maybe today, maybe tomorrow.

**Comment by Karm** [comment]
@rdebusscher I have the Java version ready, no dependencies on anything but standard javax.crypto., java.security., javax.xml., java.net.

I will draft a PR before noon. If you show me your draft of where exactly you want to produce the data, I can hook it right away to that place. If not, I post just a standalone utility class for you to use.  Note the JSON has to have a particular format defining types for the DB.

From my Java implementation demo:

```
        final String request_parameters =
                "{" +
                        "\"TableName\":\"microprofile_starter_log\"," +
                        "\"Item\":{" +
                        "\"logmark\":{\"S\":\"666\"}," +
                        "\"mpVersion\":{\"S\":\"2.0\"}," +
                        "\"supportedServer\":{\"S\":\"liberty\"}," +
                        "\"beansxmlMode\":{\"S\":\"ANNOTATED\"}," +
                        "\"moduleStructure\":{\"S\":\"SINGLE\"}," +
                        "\"selectedSpecs\":{\"SS\":[\"REST_CLIENT\",\"HEALTH_CHECKS\",\"HEALTH_METRICS\"]}," +
                        "\"timestamp\":{\"S\":\"2019-01-17T12:40:04.399+0100\"}" +
                        "}" +
                        "}";
```

Do you want me to hide that from you inside my utility class, i.e. take just attributes from you? Or would you prefer to construct the JSON yourself?

**Comment by cealsair** [comment]
+1 for adding an extra field that allows to identify came from testing or actual prod load (maybe URL)?

**Comment by Karm** [comment]
@rdebusscher See #42 that adds the utility. I will send you keys so as you can try it.

This is what it logs now:

```
$ aws dynamodb scan --table-name microprofile_starter_log 
{
    "ScannedCount": 10,
    "Items": [
        {
            "timestamp": {
                "S": "2019-01-23T12:56:13.352+0100"
            },
            "beansxmlMode": {
                "S": "ANNOTATED"
            },
            "mpVersion": {
                "S": "2.0"
            },
            "supportedServer": {
                "S": "liberty"
            },
            "selectedSpecs": {
                "SS": [
                    "HEALTH_CHECKS",
                    "HEALTH_METRICS",
                    "REST_CLIENT"
                ]
            },
            "webAppInstanceID": {
                "S": "test-instance"
            },
            "logmark": {
                "S": "379B3B1996C817C026CF2DDE6D2B628E1801E809"
            }
        },
...
...
...
    ],
    "ConsumedCapacity": null,
    "Count": 10
}
```

## To address some questions / remarks:

> That moduleStructure can be removed as not used.

Removed.

> A Thread so that posting can be asynchronous...

So far PR contains just the utility, stateless static method that could be called from your worker/runnable. If the API seems fine, I can add the integration to GeneratorDataBean.

>  hostname as a field  / instance identity

@cealsair :
There is the webAppInstanceID which is set for the container as a system property -DMP_STARTER_APP_ID   

> what is logmark?

@rdebusscher Each row has to be uniquely identified (this is a db that does not do any autoincrement something by design as it requires relation between records). Logmark is a hash of timestamp and logged data. If anybody manages to send more same download requests within the same millisecond the row gets collision and gets overwritten. I deem this not  a problem because numerous downloads with exactly same parameters including artifact and group id within a millisecond is an abuse and should not be logged as legitimate download anyway.

TL:DR: Hash. I think it is good as it is.

> add groupId and artifactId?

@cealsair  @rdebusscher 

I am using it to enhance the hash uniqueness but I do not log it due to privacy. It would potentially identify poeople's domain for their packages, their project names etc. and I do not know whether we are allowed to keep such data that would identify the user.




Good / no good?

**Comment by rdebusscher** [comment]
PR merged.
